File: convert-posts.py
#!/usr/bin/env python3
"""
Convert markdown posts to JSON format for GitHub Pages compatibility
Usage: python convert-posts.py
"""

import os
import json
import re
from pathlib import Path

class MarkdownToJSON:
    def __init__(self):
        self.posts_dir = Path('./posts')
        self.output_dir = Path('./posts-json')
        self.output_dir.mkdir(exist_ok=True)

    def extract_front_matter(self, content):
        """Extract YAML front matter from markdown content"""
        front_matter_regex = r'^---\n(.*?)\n---\n(.*)'
        match = re.match(front_matter_regex, content, re.DOTALL)
        
        if match:
            front_matter = {}
            front_matter_lines = match.group(1).strip().split('\n')
            
            for line in front_matter_lines:
                if ':' in line:
                    key, value = line.split(':', 1)
                    key = key.strip()
                    value = value.strip().strip('"\'')
                    
                    # Handle arrays (tags)
                    if value.startswith('[') and value.endswith(']'):
                        try:
                            value = json.loads(value)
                        except:
                            pass
                    
                    front_matter[key] = value
            
            return front_matter, match.group(2).strip()
        
        return {}, content

    def convert_all(self):
        """Convert all markdown files to JSON"""
        converted_posts = []
        
        # Process all markdown files
        for md_file in self.posts_dir.glob('*.md'):
            print(f"Processing: {md_file.name}")
            
            with open(md_file, 'r', encoding='utf-8') as f:
                content = f.read()
            
            front_matter, markdown_content = self.extract_front_matter(content)
            
            post_data = {
                'filename': md_file.name,
                **front_matter,
                'content': markdown_content
            }
            
            # Write individual JSON file
            json_filename = md_file.stem + '.json'
            json_path = self.output_dir / json_filename
            
            with open(json_path, 'w', encoding='utf-8') as f:
                json.dump(post_data, f, indent=2, ensure_ascii=False)
            
            converted_posts.append(post_data)
            print(f"✓ Converted: {md_file.name} → {json_filename}")
        
        # Update posts-list.json
        posts_list = []
        for post in converted_posts:
            posts_list.append({
                'filename': post['filename'].replace('.md', '.json'),
                'title': post.get('title', 'Untitled'),
                'date': post.get('date', ''),
                'category': post.get('category', 'General'),
                'excerpt': post.get('excerpt', ''),
                'tags': post.get('tags', [])
            })
        
        # Sort by date (newest first)
        posts_list.sort(key=lambda x: x['date'], reverse=True)
        
        with open('posts-list.json', 'w', encoding='utf-8') as f:
            json.dump(posts_list, f, indent=2, ensure_ascii=False)
        
        print(f"✓ Updated posts-list.json with {len(posts_list)} posts")
        return converted_posts

if __name__ == '__main__':
    converter = MarkdownToJSON()
    try:
        posts = converter.convert_all()
        print(f"🎉 Successfully converted {len(posts)} posts!")
    except Exception as e:
        print(f"❌ Error: {e}")

File: convert-posts.js
// Node.js script to convert markdown posts to JSON for GitHub Pages
const fs = require('fs');
const path = require('path');

class MarkdownToJSON {
    constructor() {
        this.postsDir = './posts';
        this.outputDir = './posts-json';
    }

    // Extract front matter from markdown
    extractFrontMatter(content) {
        const frontMatterRegex = /^---\n([\s\S]*?)\n---\n([\s\S]*)$/;
        const match = content.match(frontMatterRegex);
        
        if (match) {
            const frontMatter = {};
            const frontMatterLines = match[1].split('\n');
            
            frontMatterLines.forEach(line => {
                const [key, ...valueParts] = line.split(':');
                if (key && valueParts.length > 0) {
                    let value = valueParts.join(':').trim();

                    // Handle arrays (tags) - keep quotes for JSON parsing
                    if (value.startsWith('[') && value.endsWith(']')) {
                        value = JSON.parse(value);
                    } else {
                        // Remove quotes for regular string values
                        value = value.replace(/['"]/g, '');
                    }

                    frontMatter[key.trim()] = value;
                }
            });
            
            return {
                frontMatter,
                content: match[2]
            };
        }
        
        return {
            frontMatter: {},
            content
        };
    }

    // Convert all markdown files to JSON
    async convertAll() {
        // Create output directory
        if (!fs.existsSync(this.outputDir)) {
            fs.mkdirSync(this.outputDir);
        }

        // Read all markdown files
        const files = fs.readdirSync(this.postsDir)
            .filter(file => file.endsWith('.md'));

        const convertedPosts = [];

        for (const file of files) {
            const filePath = path.join(this.postsDir, file);
            const content = fs.readFileSync(filePath, 'utf8');
            
            const { frontMatter, content: markdownContent } = this.extractFrontMatter(content);
            
            const postData = {
                filename: file,
                ...frontMatter,
                content: markdownContent
            };

            // Write individual JSON file
            const jsonFile = file.replace('.md', '.json');
            const jsonPath = path.join(this.outputDir, jsonFile);
            fs.writeFileSync(jsonPath, JSON.stringify(postData, null, 2));
            
            convertedPosts.push(postData);
            console.log(`✓ Converted: ${file} → ${jsonFile}`);
        }

        // Update posts-list.json with the new format
        const postsList = convertedPosts.map(post => ({
            filename: post.filename.replace('.md', '.json'),
            title: post.title,
            date: post.date,
            category: post.category,
            excerpt: post.excerpt,
            tags: post.tags || []
        }));

        fs.writeFileSync('./posts-list.json', JSON.stringify(postsList, null, 2));
        console.log(`✓ Updated posts-list.json with ${postsList.length} posts`);

        return convertedPosts;
    }
}

// Run the conversion
const converter = new MarkdownToJSON();
converter.convertAll().then(() => {
    console.log('🎉 All posts converted successfully!');
}).catch(error => {
    console.error('❌ Error converting posts:', error);
});